### PR TITLE
[FEATURE][ML] Compute loss improvement in the training loop

### PR DESCRIPTION
**Pull request by tveasey**
This computes the change in training data loss in the training loop (when writing new predictions to the data frame). This will be useful to report progress down-the-line.

Also improve some naming and fix an occasional off-by-one error in training.

### CODE REVIEW

**Comment by tveasey** [comment]
retest
